[SPARK-15324] [SQL] Add the takeSample function to the Dataset

burness opened the pull request:
## What changes were proposed in this pull request?

In this pr, I add the takeSample function with the Dataset which is to sampling with the specify num instead of the fraction in sample function.
## How was this patch tested?

add a test in `DatasetSuite`

burness committed:
add takeSample in Dataset

burness committed:
modify the format of the comment

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -18,6 +18,9 @@
 import java.io.CharArrayWriter
+import java.util.Random
+
+import org.apache.spark.util.random.SamplingUtils
```
(it seems we need to reorder imports, https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-Imports)

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -402,6 +402,76 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
+    {
+      val sample = data.takeSample(withReplacement = true, num = 20)
+      assert(sample.count === 20) // Got exactly 100 elements
+    val sampleDisCount = sample.distinct.count
```
(It seems indentation is not consistent here.)

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -402,6 +402,76 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
+      val emptySet = sparkContext.parallelize(Seq.empty[Int], 2)
+      val sample = emptySet.takeSample(false, 20, 1)
+      assert(sample.length === 0)
+    }
```
(I think we might not need a extra closure here and below)

burness committed:
1, fixed the order of the imports;
2, fix the consistent of the indentation;
3, solve the extra closures by move the code to a new test case "test takeSample from a empty Dataset"

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -1494,6 +1493,56 @@ class Dataset[T] private[sql](
+      } else {
+        val rand = new Random(seed)
+        if (!withReplacement && num >= initialCount) {
+          sqlContext.createDataset[T](Utils.randomizeInPlace(this.collect(), rand))
```
This will cause memory problems (OOME) in the driver in case of a non-trivial dataset.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -1494,6 +1493,56 @@ class Dataset[T] private[sql](
+    if (num == 0) {
+      sqlContext.createDataset[T](Seq())
+    } else {
+      val initialCount = this.count()
```
This can be a non-trivial call. It will force us to evaluate a significant part of the query.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -1494,6 +1493,56 @@ class Dataset[T] private[sql](
+            withReplacement)
+          var samples = this.sample(withReplacement, fraction, rand.nextInt()).collect()
+          var numIters = 0
+          while (samples.length < num) {
```
Infinite loop if the user requests more rows than in the dataset.

hvanhovell replied:
`count` is a non-trivial action.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -1494,6 +1493,56 @@ class Dataset[T] private[sql](
+            samples = this.sample(withReplacement, fraction, rand.nextInt()).collect()
+            numIters += 1
+          }
+          sqlContext.createDataset[T](Utils.randomizeInPlace(samples, rand).take(num))
```
potential OOME

hvanhovell commented:
@burness thanks for working on this. The PR in its current state has some serious potential memory and performance problems (see the comments). What is the usecase for this? If you do want to press ahead you will need to take a look at the physical `SampleExec` operator and somekind of limit to that.

burness commented:
@hvanhovell  Thank you for code review. In my project, I want to sample by the specify num but in DataFrame or Dataset, there is only the sample by fraction. And I have take these days , but it's not easy for me to resolve the serious potential memory and performance problems(only add the line to require that the sample num shouldn't bigger than the dataset's size).

hvanhovell commented:
@burness that is no problem. I have taken a look at the current implementation of the `RDD`'s `takeSample`; it is very similar to your implementation. I do still think that implementation has problems, and I am not sure we should add to the `Dataset`/`Dataframe` API. 

You can always use the datasets 's RDD use its `takeSample` method, i.e.: `spark.createDataset(dataset.rdd.takeSample(...))`.

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -51,6 +49,7 @@ import org.apache.spark.sql.execution.python.EvaluatePython
 import org.apache.spark.sql.types._
 import org.apache.spark.storage.StorageLevel
 import org.apache.spark.util.Utils
+import org.apache.spark.util.random.SamplingUtils
```
I think the imports might be able to be as below:

```scala

import java.io.CharArrayWriter
import java.util.Random

import scala.collection.JavaConverters._
import scala.language.implicitConversions
import scala.reflect.runtime.universe.TypeTag
import scala.util.control.NonFatal

import com.fasterxml.jackson.core.JsonFactory
import org.apache.commons.lang3.StringUtils

import org.apache.spark.annotation.{DeveloperApi, Experimental}
import org.apache.spark.api.java.JavaRDD
import org.apache.spark.api.java.function._
import org.apache.spark.api.python.PythonRDD
import org.apache.spark.broadcast.Broadcast
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.catalyst._
import org.apache.spark.sql.catalyst.analysis._
import org.apache.spark.sql.catalyst.encoders._
import org.apache.spark.sql.catalyst.expressions._
import org.apache.spark.sql.catalyst.expressions.aggregate._
import org.apache.spark.sql.catalyst.expressions.objects.Invoke
import org.apache.spark.sql.catalyst.optimizer.CombineUnions
import org.apache.spark.sql.catalyst.plans._
import org.apache.spark.sql.catalyst.plans.logical._
import org.apache.spark.sql.catalyst.util.usePrettyExpression
import org.apache.spark.sql.execution.{FileRelation, LogicalRDD, QueryExecution, SQLExecution}
import org.apache.spark.sql.execution.command.ExplainCommand
import org.apache.spark.sql.execution.datasources.{CreateTableUsingAsSelect, LogicalRelation}
import org.apache.spark.sql.execution.datasources.json.JacksonGenerator
import org.apache.spark.sql.execution.python.EvaluatePython
import org.apache.spark.sql.types._
import org.apache.spark.storage.StorageLevel
import org.apache.spark.util.Utils
import org.apache.spark.util.random.SamplingUtils
```

burness commented:
@hvanhovell yeah, I agree with your opinion！ Thanks

HyukjinKwon commented:
Hi @burness, what's the state of this PR?

burness commented:
@HyukjinKwon  It is too hard to solve the OOM, I'm so sorry

HyukjinKwon commented:
Ah, that's fine. I just wondered if this is on a progress and we maybe close this for now.

burness commented:
@HyukjinKwon  Ok, I will close this now 😆
